### ISSUE TITLE
Problem with dark editor theme

### ISSUE THREAD

**Issue by asampal** [laurentpetit/ccw#799]
If the editor background color is set to black, it remains white except for the line number columns.

Eclipse 4.5, Windows 8.1
ccw.core 0.32.0.201506262242

**Comment by laurentpetit** [comment]
Reproduced.
@arichiardi do you think you'll have the possibiity to take a look, or at least point me in the right direction?

**Comment by arichiardi** [comment]
I think yes i will try this weekend, it should not be some issue with tokens, probably the background pref not read/set correctly

**Comment by laurentpetit** [comment]
Hello, 




I have found the time to look at it and I think I have a solution. I will submit a PR for review, if you're ok with this?



—
Envoyé avec Mailbox

**Comment by arichiardi** [comment]
Oh yes please go ahead,  i did not have time to
 have a look but i will test it for sure

**Comment by laurentpetit** [comment]
@arichiardi the fix is in this commit, can you review it? https://github.com/laurentpetit/ccw/commit/6fd012a38e6fa3ee3b90b9cc78beb49d22385124

**Comment by asampal** [comment]
The fix works, but there is still a one character wide column that remains white, separating the line number area from the editing area.

**Comment by laurentpetit** [comment]
@asampal indeed, will work on this

**Comment by arichiardi** [comment]
After having debugged ClojureTokenScanner, I discovered that it was not related...
Probably folding is the culprit...https://stackoverflow.com/questions/12207637/eclipse-juno-how-to-change-background-of-code-folding-regions

**Comment by arichiardi** [comment]
Another related issue:
https://stackoverflow.com/questions/12490672/vertical-white-line-on-eclipse

@asampal could you try these two workarounds?

**Comment by arichiardi** [comment]
And finally:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=194313

**Comment by laurentpetit** [comment]
Yes, folding seems to be the issue. JDT does not have the problem.

**Comment by arichiardi** [comment]
Sorry I have posted a workaround for CDT, will try something better, for example I can try emulating what JDT is doing

**Comment by asampal** [comment]
Toggling the folding functionality didn't fix this for me. Closed the editors and re-opened, but no go.

**Comment by laurentpetit** [comment]
@arichiardi I'd like to focus on what's required to release a new stable version. This issue seems like an embarrassing regression to me. Better to fix it soon. Can you give it another try?

**Comment by arichiardi** [comment]
Yes ok I will focus on this then

**Comment by laurentpetit** [comment]
thanks a lot!

**Comment by arichiardi** [comment]
Don't mention it :smile: !

**Comment by laurentpetit** [comment]
Just to refresh my memory, I gave the current stable version a try, and it does not have the problem, indeed.
I tried to disable as many options as possible related to folding, but the problem remains the same, so it may not be related to the same problem people have reported on Stack Overflow.

Will try to bisect to find where the problem occurs.

**Comment by laurentpetit** [comment]
OK, so bisecting confirms that fcc360c is the first commit introducing the problem

**Comment by laurentpetit** [comment]
Well, big commit, hard to tell where the problem lies. Will try another approach and put some breakpoints in the StyledText

**Comment by arichiardi** [comment]
Will debug too, I remember I did change nothing about the tokens...it is weird because it looks like that stripe is not even within buffer range...I will help you out and sorry about it..

**Comment by arichiardi** [comment]
initializeViewerColors looks suspicious for instance, maybe it does not take the black background color

**Comment by arichiardi** [comment]
Another road I am considering: https://github.com/guari/eclipse-ui-theme/issues/7

**Comment by arichiardi** [comment]
I put here the SWT spy log:
```
AnnotationRulerColumn$5 {}@140230138540928
	Style: NO_FOCUS | LEFT_TO_RIGHT | DOUBLE_BUFFERED 
	Layout Data: null
	Bounds: Rectangle {44, 0, 9, 206}


Children:

Peers:
	*AnnotationRulerColumn$5 {}@140230138540928 Layout Data: null Bounds: Rectangle {44, 0, 9, 206}
	AnnotationRulerColumn$5 {}@140230141853568 Layout Data: null Bounds: Rectangle {0, 0, 12, 206}
	LineNumberRulerColumn$4 {}@140230140489248 Layout Data: null Bounds: Rectangle {12, 0, 32, 206}

Parent Tree:
[CUT]
```

**Comment by arichiardi** [comment]
I am digging deep, basically the setBackground that follows correctly shows {255,0,0} the red I chose:
![selection_024](https://cloud.githubusercontent.com/assets/1724788/8953630/314df002-35e0-11e5-9fbf-52349222e350.png)

But when ```AnnotationRulerColumn.redraw()``` is triggered, the background color on its ```fCanvas``` is {215,215,215}, which is that light gray you see...

Now, why in the world the color goes in the ```StyledText``` and not in the ```AnnotationRulerColumn``` I don't know...

**Comment by laurentpetit** [comment]
Yeah, I also already spent a night on this yesterday without success. But I'm stubborn, and I hope we'll find the problem. We may also learn things about how swt / jface / viewers work, along the way. 



—
Envoyé avec Mailbox

**Comment by arichiardi** [comment]
By the way it still happens in my Mars with jdt...have you tried to put a color that contrasts your theme color? We'll better report this in [here](https://bugs.eclipse.org/bugs/show_bug.cgi?id=194313).

![selection_025](https://cloud.githubusercontent.com/assets/1724788/8964523/5a113f08-3626-11e5-8785-9c2d611ade38.png)

**Comment by laurentpetit** [comment]
Well, on my Mars, java editor does not seem to have problems, while clojure editor does.

**Comment by arichiardi** [comment]
Could also be OS related, or Eclipse related, I am running on linux and y target platform is kepler atm.

**Comment by laurentpetit** [comment]
Indeed. I think I can probably release with this problem anyway. It will
not be too late to fix it later, and it's kind of a shame that I'm holding
back all the good stuff that's been done since January !

2015-07-29 21:01 GMT+02:00 Andrea Richiardi <notifications@github.com>:

> Could also be OS related, or Eclipse related, I am running on linux and y
> target platform is kepler atm.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laurentpetit/ccw/issues/799#issuecomment-126060016>.
>



-- 
Laurent Petit

**Comment by laurentpetit** [comment]
Here's what I have here, on OS X, running Mars :
![screenshot 2015-07-29 21 15 05](https://cloud.githubusercontent.com/assets/98286/8967427/1293fb8c-3637-11e5-96cf-60f5ec81774c.png)

**Comment by arichiardi** [comment]
That's true ;)

On July 29, 2015 9:14:10 PM GMT+02:00, Laurent Petit <notifications@github.com> wrote:
>Indeed. I think I can probably release with this problem anyway. It
>will
>not be too late to fix it later, and it's kind of a shame that I'm
>holding
>back all the good stuff that's been done since January !
>
>2015-07-29 21:01 GMT+02:00 Andrea Richiardi <notifications@github.com>:
>
>> Could also be OS related, or Eclipse related, I am running on linux
>and y
>> target platform is kepler atm.
>>
>> —
>> Reply to this email directly or view it on GitHub
>>
><https://github.com/laurentpetit/ccw/issues/799#issuecomment-126060016>.
>>
>
>
>
>-- 
>Laurent Petit
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/laurentpetit/ccw/issues/799#issuecomment-126063034

-ar

**Comment by arichiardi** [comment]
Even with folding, this problem is still there...very weird.

**Comment by arichiardi** [comment]
I mean in my branch folding-feature

**Comment by arichiardi** [comment]
I have randomly tried commit ```ae2e517``` and it still has the problem, older than fcc360c88dfe5918681321a91c5eb2595eb4182e

**Comment by arichiardi** [comment]
In my Mars with Color Theme installed I went to Text Editors and reset to default colors...and...

![selection_002](https://cloud.githubusercontent.com/assets/1724788/9488231/98bd1caa-4bd8-11e5-82b9-def632aa929c.png)

It looks like this bug is hidden in JDT, but it is still there...but still, we should try to fix it in CCW first.

**Comment by arichiardi** [comment]
Good news! I solved the problem!

![selection_003](https://cloud.githubusercontent.com/assets/1724788/9527955/65a2ef78-4cf2-11e5-9368-2fede2a08699.png)

**Comment by arichiardi** [comment]
It will be in thn folding-feature because it is directly related with it, difficult to extrapolate...

**Comment by laurentpetit** [comment]
Good news indeed! What was the problem?

**Comment by arichiardi** [comment]
When you change the background basically, because of the above-mentioned bug, you need to toggle the folding on PrjectionViewer.
I am planning to do it automagically thanks to your whiteboard pattern and listening to the background preference.

**Comment by laurentpetit** [comment]
(y)

**Comment by arichiardi** [comment]
Actually just listening to the background should suffice..
